### DISCUSSION TITLE
Edit data of all messages (value) before submitting to DLT topic

### DISCUSSION THREAD

**Discussion by Watlas** [spring-projects/spring-kafka#2265] [Q&A]
Hello, I was reading the documentation and could not identify a solution to the following problem:

I want to add new attributes to the message before sending to the DLT topic, injecting this at the configuration level, I want to add attributes to the message body coming from the exception itself.

I believe what I'm looking for is a type of interceptor that allows me to change data.

**Comment by artembilan** [comment]
By default the framework uses a `DeadLetterPublishingRecoverer`. This one has a method like:
```
	/**
	 * Subclasses can override this method to customize the producer record to send to the
	 * DLQ. The default implementation simply copies the key and value from the consumer
	 * record and adds the headers. The timestamp is not set (the original timestamp is in
	 * one of the headers). IMPORTANT: if the partition in the {@link TopicPartition} is
	 * less than 0, it must be set to null in the {@link ProducerRecord}.
	 * @param record the failed record
	 * @param topicPartition the {@link TopicPartition} returned by the destination
	 * resolver.
	 * @param headers the headers - original record headers plus DLT headers.
	 * @param key the key to use instead of the consumer record key.
	 * @param value the value to use instead of the consumer record value.
	 * @return the producer record to send.
	 * @see KafkaHeaders
	 */
	protected ProducerRecord<Object, Object> createProducerRecord(ConsumerRecord<?, ?> record,
			TopicPartition topicPartition, Headers headers, @Nullable byte[] key, @Nullable byte[] value) {

		return new ProducerRecord<>(topicPartition.topic(),
				topicPartition.partition() < 0 ? null : topicPartition.partition(),
				key != null ? key : record.key(),
				value != null ? value : record.value(), headers);
	}
```

so, you can extend this  `DeadLetterPublishingRecoverer` and override that method whatever way you need.


You inject your custom recoverer into a `DefaultErrorHandler`: https://docs.spring.io/spring-kafka/docs/current/reference/html/#dead-letters.
Which you will provide for the listener container factory:
```
	/**
	 * Set the {@link CommonErrorHandler} which can handle errors for both record and
	 * batch listeners. Replaces the use of
	 * {@link org.springframework.kafka.listener.GenericErrorHandler}s.
	 * @param commonErrorHandler the handler.
	 * @since 2.8
	 */
	public void setCommonErrorHandler(CommonErrorHandler commonErrorHandler) {
```

**Reply by Watlas** [reply]
my custom class:

![image](https://user-images.githubusercontent.com/57519850/167189662-fe64e577-526b-4dd2-b978-fd90be3e23ce.png)

my injection:
![image](https://user-images.githubusercontent.com/57519850/167190039-9334ff32-60f7-49e1-881a-7f21190235db.png)


can you tell me what i can be doing wrong?

**Reply by artembilan** [reply]
Sorry, you didn't say what's error is to give me more context of the problem.

The code looks OK for me, unless you just don't do any `ProducerRecord` customization as you mentioned originally...

**Reply by Watlas** [reply]
I apologize for not specifying the error.

There is no error, it simply does not enter the method:
```
@Override
createProducerRecord
```
This log I put was just to know if he was running the method or not.



```
    @Override
    protected ProducerRecord<Object, Object> createProducerRecord(ConsumerRecord<?, ?> record, TopicPartition topicPartition, Headers headers, byte[] key, byte[] value) {
        record = new ConsumerRecord<Object, Object>(record.topic(), record.partition(),
                record.partition(), record.timestamp(), record.timestampType(),
                record.serializedKeySize(), record.serializedValueSize(), record.key(),
                "teste dql", record.headers(), record.leaderEpoch());
        return super.createProducerRecord(record, topicPartition, headers, key, value);
    }
```

even if I change it, it doesn't fall into the method

**Reply by artembilan** [reply]
Well. now we need to understand how you use this `kafkaListenerContainerFactory`.

Anyway, you always can debug open source code to see where it leads when you consume a record from the topic.

**Reply by Watlas** [reply]
It worked well, but there was another problem, when I try to join this solution with the RetryTopicConfiguration it stops working, could you explain why?

![image](https://user-images.githubusercontent.com/57519850/167228865-cfe8824a-ddee-4318-96a6-ec112af4d259.png)

**Reply by artembilan** [reply]
Not yet. Will look on Monday.
Perhaps @tomazfernandes can spot something immediately.
Thanks

**Reply by tomazfernandes** [reply]
I don't think the [DeadLetterPublishingRecovererFactory](https://github.com/spring-projects/spring-kafka/blob/731af3223aa47891cb1a3a4a3eb6dfb8b3a338a2/spring-kafka/src/main/java/org/springframework/kafka/retrytopic/DeadLetterPublishingRecovererFactory.java) currently offers any extension points that would enable replacing the record, or using a `DeadLetterPublishingFactory` subclass instead of the provided one.

You can open an issue so we can look into offering one or both of these.

Thanks.

**Reply by Watlas** [reply]
@tomazfernandes So it is currently not possible to use the non-blocking retry solution and the ProducerRecord edit solution together?

**Reply by garyrussell** [reply]
Correct. We need to refactor the `DeadLetterPublishingRecovererFactory` to allow creation of the subclass. Not too difficult, but it will require some other refactoring because the non-blocking retry mechanism uses different header names.

**Reply by tomazfernandes** [reply]
Agreed. Another option would be extracting the anonymous retry topic `DLPR` subclass into a proper class, so that users could subclass that instead, and then just offer an extensibility point for the actual instance creation.

For the matter at hand, we do offer extensibility in `ListenerContainerFactoryConfigurer#createDefaultErrorHandlerInstance`, so we could replace the `DefaultErrorHandler` with a subclass. If we can intercept the `ConsumerRecord` instead to change the payload, we might be able to achieve the expected result. Not an ideal solution of course.
